Mathlib/Tactic/ApplyFun.lean

```diff
@@ -24,32 +24,64 @@ initialize registerTraceClass `apply_fun
+    if ty.isForall then
+      if ty.bindingInfo! == .implicit then
+        let mvar ← mkFreshExprMVar ty.bindingDomain! MetavarKind.natural ty.bindingName!
+        return ← loop (mvars.push mvar.mvarId!) (ty.bindingBody!.instantiate1 mvar) (e.beta #[mvar])
+      else if ty.bindingInfo! == .instImplicit then
+        let mvar ← mkFreshExprMVar ty.bindingDomain! MetavarKind.synthetic ty.bindingName!
+        return ← loop (mvars.push mvar.mvarId!) (ty.bindingBody!.instantiate1 mvar) (e.beta #[mvar])
+    return (e, mvars)
+
+/-- Returns a function. Ensures all leading implicit variables are filled, and (recursively)
+uses `CoeFun` instances. -/
+private def ensureFun (f : Expr) (steps : Nat := 10000) : MetaM (Expr × Array MVarId) :=
+  match steps with
+  | 0 => throwError "Could not coerce to function (iteration limit reached)"
+  | steps + 1 => do
+    let (f, mvars) ← fillImplicitArgumentsWithFreshMVars f
+    if (← inferType f).isForall then
+      return (f, mvars)
+    else
+      let some f ← coerceToFunction? f
+        | throwError "Could not coerce to function"
+      let (f', mvars') ← ensureFun f steps
+      return (f', mvars ++ mvars')
 
 /-- Apply a function to a hypothesis. -/
 def applyFunHyp (f : Expr) (using? : Option Expr) (h : FVarId) (g : MVarId) :
     MetaM (List MVarId) := do
   let d ← h.getDecl
   let (prf, newGoals) ← match (← whnfR (← instantiateMVars d.type)).getAppFnArgs with
-  | (``Eq, #[α, _, _]) => do
-    -- We have to jump through a hoop here!
-    -- At this point Lean may think `f` is a dependently-typed function,
-    -- so we can't just feed it to `congrArg`.
-    -- To solve this, we first fill any implicit arguments for `f` with metavariables,
-    -- and then try to unify with a metavariable `_ : α → _`
-    -- (i.e. an arrow, but with the target as some fresh type metavariable).
-    -- (Arguably `Lean.Meta.mkCongrArg` could do this all itself.)
-    let arrow ← mkFreshExprMVar (← mkArrow α (← mkFreshTypeMVar))
-    let f' ← fillImplicitArgumentsWithFreshMVars f
-    if ¬ (← isDefEq f' arrow) then
-      throwError "Can not use `apply_fun` with a dependently typed function."
-    pure (← mkCongrArg f' d.toExpr, [])
+  | (``Eq, #[ty, lhs, rhs]) => do
+    let (f, mvars) ← ensureFun f
+    let argTy := (← inferType f).bindingDomain!
+    unless ← isDefEq argTy ty do
+      throwAppTypeMismatch argTy ty
+    -- Note: there might be implicit arguments *after* lhs and rhs
+    let (lhs', mvarslhs) ← fillImplicitArgumentsWithFreshMVars (f.beta #[lhs])
+    let (rhs', mvarsrhs) ← fillImplicitArgumentsWithFreshMVars (f.beta #[rhs])
+    unless ← isDefEq (← inferType lhs') (← inferType rhs') do
+      let msg ← mkHasTypeButIsExpectedMsg (← inferType rhs') (← inferType lhs')
+      throwError "In generated equality, right-hand side {msg}"
+    let mvars := mvars ++ mvarslhs ++ mvarsrhs
+    -- `mkAppN` would do this, but it uses `withNewMCtxDepth`.
```

Comment:
Could you explain the problem here?

Reply:
There are sort of two problems:

1. `mkAppN` (because it uses `withNewMCtxDepth`) won't assign any metavariables, but we're wanting to solve for the ones we introduce from the whole `ensureFun` algorithm.
2. `mkAppN` checks that there are no assignable metavariables at the end

I know there are functions like `withCollectingNewGoalsFrom`, and I imagine the `TermElab` monad could be helpful, though I wasn't sure how to wire these things up here to help construct expressions, collect metavariables, then synthesize them before returning.

Reply:
Another thing that's surprising is that in the test with `apply_fun (fun n => n+1) at h` the generated `h` has a whole lot of metavariables (it hasn't decided about the types of anything in the `hAdd` but the `n`), so it wouldn't pass `mkAppN`'s metavariable check anyway.

And I guess because we're using `elabTermForApply` we don't immediately get a list of unassigned metavariables, so there are no guarantees that these metavariables will ever be accounted for.